File: Data Structures & Algorithms/jump-game-ii/submission-1.java
class Solution {
    public int jump(int[] nums) {
        return dfs(nums, 0);
    }

    private int dfs(int[] nums, int i) {
        if (i == nums.length - 1) {
            return 0;
        }
        if (nums[i] == 0) {
            return 1000000;
        }
        int res = 1000000;
        int end = Math.min(nums.length - 1, i + nums[i]);
        for (int j = i + 1; j <= end; j++) {
            res = Math.min(res, 1 + dfs(nums, j));
        }
        return res;
    }
}

File: Data Structures & Algorithms/maximum-profit-in-job-scheduling/submission-1.cs
public class Solution {
    public int JobScheduling(int[] startTime, int[] endTime, int[] profit) {
        int n = startTime.Length;
        var intervals = new List<(int start, int end, int profit)>();

        for (int i = 0; i < n; i++) {
            intervals.Add((startTime[i], endTime[i], profit[i]));
        }

        intervals.Sort((a, b) => a.start.CompareTo(b.start));
        Dictionary<int, int> cache = new();

        int Dfs(int i) {
            if (i == n) return 0;
            if (cache.ContainsKey(i)) return cache[i];

            // Option 1: don't include
            int res = Dfs(i + 1);

            // Option 2: include current job
            int j = i + 1;
            while (j < n && intervals[j].start < intervals[i].end) {
                j++;
            }

            res = Math.Max(res, intervals[i].profit + Dfs(j));
            cache[i] = res;
            return res;
        }

        return Dfs(0);
    }
}